packages/svelte/src/internal/client/runtime.js

```diff
@@ -228,15 +229,63 @@ export function check_dirtiness(reaction) {
 	return false;
 }
 
+/**
+ * @param {Error} error
+ * @param {Effect} effect
+ */
+function propagate_error(error, effect) {
+	/** @type {Effect | null} */
+	var current = effect;
+	while (current !== null) {
+		/** @type {Effect | null} */
+		var parent = current.parent;
+
+		if ((current.f & BOUNDARY_EFFECT) !== 0) {
+			try {
+				// @ts-ignore
+				current.fn(error);
+			} catch {
+				// Remove boundary flag from effect
+				current.f ^= BOUNDARY_EFFECT;
+				current = parent;
+				continue;
+			}
+			return;
+		}
+		current = parent;
+	}
+
+	throw error;
+}
+
+/**
+ * @param {Effect} effect
+ */
+function should_rethrow_error(effect) {
+	return (
+		(effect.f & DESTROYED) === 0 &&
+		(effect.parent === null || (effect.parent.f & BOUNDARY_EFFECT) === 0)
+	);
+}
+
 /**
  * @param {Error} error
  * @param {Effect} effect
  * @param {ComponentContext | null} component_context
  */
-function handle_error(error, effect, component_context) {
-	// Given we don't yet have error boundaries, we will just always throw.
-	if (!DEV || handled_errors.has(error) || component_context === null) {
-		throw error;
+export function handle_error(error, effect, component_context) {
+	if (handled_errors.has(error)) {
+		if (should_rethrow_error(effect)) {
+			throw error;
+		}
+		return;
+	}
+
+	handled_errors.add(error);
```

Comment:
Technically throwing non-error objects still works in this scenario, which means that typing `error` as an `Error` is strictly not correct. The react `componentDidCatch` lifecycle hook used to do this, and it's fixing it's types to be `unknown` in it's next major: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69434.

If you do something like `throw 123;` you end up getting an error thrown by this code:

```
WeakSet value 123 must be an object
  
  in $effect.pre
  in ChildThird.svelte
  in ChildSecond.svelte
  in ChildFirst.svelte
  in App.svelte

handle_error@playground:output:1585:18 update_effect@playground:output:1804:16 process_effects@playground:output:1980:22 flush_queued_root_effects@playground:output:1863:20 process_deferred@playground:output:1916:28 handle_error@playground:output:1585:18 update_effect@playground:output:1804:16 process_effects@playground:output:1980:22 flush_queued_root_effects@playground:output:1863:20 process_deferred@playground:output:1916:28
```

should we try to guard against this here? It also means that accessing error properties in `onerror` might causing more errors to be thrown.

If we do end up guarding it here, we might have to make `onerror` take `unknown` instead of `Error` as it's first argument. This is something I PR'd into React earlier this year, it's going to be fixed with the next React major: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69434

Reply:
Good point. We now only handle errors only and with anything else we re-throw them.

Reply:
I think that's the wrong conclusion. If something is thrown we catch it, no exceptions (ha). So I'd rather adjust the catch logic to not have a stack in that case and also adjust the types to `unknown`

Reply:
It’s a bit more complicated than that we use the WeakSet to avoid re-applying propagation on the error.

Reply:
We could allow objects to be thrown with this, but I can't see any way we can make this work with primitives. I considered using a Map, but there's no ideal place to clear the Map without possibly causing other problems?

Reply:
Actually I have another approach, will push that now.

Reply:
Just as a thought: What if we wrap non-errors in an error object and put the original on some property? Or does that mess with user expectations too much? Not sure if doing that would help you have nicer/smaller logic in the code. It would at least help with attaching a stack, though that's probably not the most important thing.

Reply:
I considered that, but it had the unfortunate side-effect of us needing to ensure we unwrapped the error again in all the different places that it can escape. So it ended up adding just as much code.